[bug] ruff chokes when fixing invalid comparison (`F632`)

This `F632`-triggering line (and several similar ones in our codebase) caused `ruff --fix` to choke:

```
type_check_fn=lambda _, value: isinstance(value, int) and value % 2 is 0,
```

Error message:

```
[2022-12-05][08:09:02][ruff::pyflakes::plugins::invalid_literal_comparisons][ERROR] Failed to fix invalid comparison: Failed to extract CST from source
```

Subsequent runs of `ruff --fix` print no error message but don't apply the fix either (though I think this is the cache? Running with `--no-cache` prints the error again).

--- a/resources/test/fixtures/pyflakes/F632.py
+++ b/resources/test/fixtures/pyflakes/F632.py
@@ -4,5 +4,18 @@ if x is "abc":
 if 123 is not y:
     pass
 
+if 123 is \
+        not y:
+    pass
+
 if "123" is x < 3:
+    pass
+
+if "123" != x is 3:
+    pass
+
+if ("123" != x) is 3:
+    pass
+
+if "123" != (x is 3):
     pass

--- a/src/ast/operations.rs
+++ b/src/ast/operations.rs
@@ -1,4 +1,7 @@
+use rustpython_ast::{Cmpop, Located};
 use rustpython_parser::ast::{Constant, Expr, ExprKind, Stmt, StmtKind};
+use rustpython_parser::lexer;
+use rustpython_parser::lexer::Tok;
 
 use crate::ast::types::{BindingKind, Scope};
 
@@ -108,4 +111,149 @@ pub fn is_unpacking_assignment(stmt: &Stmt) -> bool {
         &value.node,
         ExprKind::Set { .. } | ExprKind::List { .. } | ExprKind::Tuple { .. }
     )
+}
+
+pub type LocatedCmpop<U = ()> = Located<Cmpop, U>;
+
+/// Extract all `Cmpop` operators from a source code snippet, with appropriate
+/// ranges.
+///
+/// RustPython doesn't include line and column information on `Cmpop` nodes.
+/// (CPython doesn't either.) This method iterates over the token stream and
+/// re-identifies `Cmpop` nodes, annotating them with valid arnges.
+pub fn locate_cmpops(contents: &str) -> Vec<LocatedCmpop> {
+    let mut tok_iter = lexer::make_tokenizer(contents)
+        .flatten()
+        .into_iter()
+        .peekable();
+    let mut ops: Vec<LocatedCmpop> = vec![];
+    let mut count: usize = 0;
+    loop {
+        let Some((start, tok, end)) = tok_iter.next() else {
+            break;
+        };
+        if matches!(tok, Tok::Lpar) {
+            count += 1;
+            continue;
+        } else if matches!(tok, Tok::Rpar) {
+            count -= 1;
+            continue;
+        }
+        if count == 0 {
+            match tok {
+                Tok::Not => {
+                    if let Some((_, _, end)) =
+                        tok_iter.next_if(|(_, tok, _)| matches!(tok, Tok::In))
+                    {
+                        ops.push(LocatedCmpop::new(start, end, Cmpop::NotIn));
+                    }
+                }
+                Tok::In => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::In));
+                }
+                Tok::Is => {
+                    if let Some((_, _, end)) =
+                        tok_iter.next_if(|(_, tok, _)| matches!(tok, Tok::Not))
+                    {
+                        ops.push(LocatedCmpop::new(start, end, Cmpop::IsNot));
+                    } else {
+                        ops.push(LocatedCmpop::new(start, end, Cmpop::Is));
+                    }
+                }
+                Tok::NotEqual => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::NotEq));
+                }
+                Tok::EqEqual => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::Eq));
+                }
+                Tok::GreaterEqual => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::GtE));
+                }
+                Tok::Greater => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::Gt));
+                }
+                Tok::LessEqual => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::LtE));
+                }
+                Tok::Less => {
+                    ops.push(LocatedCmpop::new(start, end, Cmpop::Lt));
+                }
+                _ => {}
+            }
+        }
+    }
+    ops
+}
+
+#[cfg(test)]
+mod tests {
+    use rustpython_ast::{Cmpop, Location};
+
+    use crate::ast::operations::{locate_cmpops, LocatedCmpop};
+
+    #[test]
+    fn locates_cmpops() {
+        assert_eq!(
+            locate_cmpops("x == 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 4),
+                Cmpop::Eq
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x != 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 4),
+                Cmpop::NotEq
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x is 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 4),
+                Cmpop::Is
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x is not 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 8),
+                Cmpop::IsNot
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x in 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 4),
+                Cmpop::In
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x not in 1"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 8),
+                Cmpop::NotIn
+            )]
+        );
+
+        assert_eq!(
+            locate_cmpops("x != (1 is not 2)"),
+            vec![LocatedCmpop::new(
+                Location::new(1, 2),
+                Location::new(1, 4),
+                Cmpop::NotEq
+            )]
+        );
+    }
 }

--- a/src/pyflakes/fixes.rs
+++ b/src/pyflakes/fixes.rs
@@ -1,14 +1,11 @@
 use anyhow::{bail, Result};
-use libcst_native::{
-    Codegen, CodegenState, CompOp, Comparison, ComparisonTarget, Expr, Expression, ImportNames,
-    SmallStatement, Statement,
-};
+use libcst_native::{Codegen, CodegenState, ImportNames, SmallStatement, Statement};
 use rustpython_ast::Stmt;
 
 use crate::ast::types::Range;
 use crate::autofix::{helpers, Fix};
 use crate::cst::helpers::compose_module_path;
-use crate::cst::matchers::{match_expr, match_module};
+use crate::cst::matchers::match_module;
 use crate::source_code_locator::SourceCodeLocator;
 
 /// Generate a Fix to remove any unused imports from an `import` statement.
@@ -75,62 +72,4 @@ pub fn remove_unused_imports(
             stmt.end_location.unwrap(),
         ))
     }
-}
-
-fn match_comparison<'a, 'b>(expr: &'a mut Expr<'b>) -> Result<&'a mut Comparison<'b>> {
-    if let Expression::Comparison(comparison) = &mut expr.value {
-        Ok(comparison)
-    } else {
-        bail!("Expected Expression::Comparison")
-    }
-}
-
-/// Generate a Fix to replace invalid is/is not comparisons with equal/not equal
-pub fn fix_invalid_literal_comparison(locator: &SourceCodeLocator, location: Range) -> Result<Fix> {
-    let module_text = locator.slice_source_code_range(&location);
-    let mut tree = match_module(&module_text)?;
-    let mut expr = match_expr(&mut tree)?;
-    let cmp = match_comparison(expr)?;
-    let target = cmp
-        .comparisons
-        .get(0)
-        .ok_or_else(|| anyhow::anyhow!("Expected one ComparisonTarget"))?;
-
-    let new_operator = match &target.operator {
-        CompOp::Is {
-            whitespace_before: b,
-            whitespace_after: a,
-        } => CompOp::Equal {
-            whitespace_before: b.clone(),
-            whitespace_after: a.clone(),
-        },
-        CompOp::IsNot {
-            whitespace_before: b,
-            whitespace_after: a,
-            whitespace_between: _,
-        } => CompOp::NotEqual {
-            whitespace_before: b.clone(),
-            whitespace_after: a.clone(),
-        },
-        op => bail!("Unexpected operator: {op:?} (expected CompOp::Is or CompOp::IsNot)"),
-    };
-
-    expr.value = Expression::Comparison(Box::new(Comparison {
-        left: cmp.left.clone(),
-        comparisons: vec![ComparisonTarget {
-            operator: new_operator,
-            comparator: target.comparator.clone(),
-        }],
-        lpar: cmp.lpar.clone(),
-        rpar: cmp.rpar.clone(),
-    }));
-
-    let mut state = CodegenState::default();
-    tree.codegen(&mut state);
-
-    Ok(Fix::replacement(
-        state.to_string(),
-        location.location,
-        location.end_location,
-    ))
 }

--- a/src/pyflakes/plugins/invalid_literal_comparisons.rs
+++ b/src/pyflakes/plugins/invalid_literal_comparisons.rs
@@ -1,11 +1,13 @@
 use itertools::izip;
-use log::error;
+use once_cell::unsync::Lazy;
 use rustpython_ast::{Cmpop, Constant, Expr, ExprKind};
 
+use crate::ast::helpers;
+use crate::ast::operations::locate_cmpops;
 use crate::ast::types::Range;
+use crate::autofix::Fix;
 use crate::check_ast::Checker;
 use crate::checks::{Check, CheckKind};
-use crate::pyflakes::fixes::fix_invalid_literal_comparison;
 
 fn is_singleton(expr: &Expr) -> bool {
     matches!(
@@ -37,22 +39,35 @@ pub fn invalid_literal_comparison(
     comparators: &[Expr],
     location: Range,
 ) {
+    let located = Lazy::new(|| locate_cmpops(&checker.locator.slice_source_code_range(&location)));
     let mut left = left;
-    for (op, right) in izip!(ops, comparators) {
+    for (index, (op, right)) in izip!(ops, comparators).enumerate() {
         if matches!(op, Cmpop::Is | Cmpop::IsNot)
             && (is_constant_non_singleton(left) || is_constant_non_singleton(right))
         {
             let mut check = Check::new(CheckKind::IsLiteral, location);
             if checker.patch(check.kind.code()) {
-                match fix_invalid_literal_comparison(
-                    checker.locator,
-                    Range {
-                        location: left.location,
-                        end_location: right.end_location.unwrap(),
-                    },
-                ) {
-                    Ok(fix) => check.amend(fix),
-                    Err(e) => error!("Failed to fix invalid comparison: {e}"),
+                if let Some(located_op) = &located.get(index) {
+                    assert_eq!(&located_op.node, op);
+                    if let Some(content) = match &located_op.node {
+                        Cmpop::Is => Some("==".to_string()),
+                        Cmpop::IsNot => Some("!=".to_string()),
+                        node => {
+                            eprintln!("Failed to fix invalid comparison: {node:?}");
+                            None
+                        }
+                    } {
+                        check.amend(Fix::replacement(
+                            content,
+                            helpers::to_absolute(located_op.location, location.location),
+                            helpers::to_absolute(
+                                located_op.end_location.unwrap(),
+                                location.location,
+                            ),
+                        ));
+                    }
+                } else {
+                    eprintln!("Failed to fix invalid comparison due to missing op");
                 }
             }
             checker.add_check(check);

